grin/main.py

```diff
@@ -92,15 +92,21 @@ def read_block_with_context(self, prev, fp, fp_size):
         -------
         A DataBlock representing the "current" block along with context.
         """
-        if fp_size is None:
+        try:
+            if fp_size is None:
+                target_io_size = READ_BLOCKSIZE
+                block_main = fp.read(target_io_size)
+                is_last_block = len(block_main) < target_io_size
+            else:
+                remaining = max(fp_size - fp.tell(), 0)
+                target_io_size = min(READ_BLOCKSIZE, remaining)
+                block_main = fp.read(target_io_size)
+                is_last_block = target_io_size == remaining
+        except UnicodeDecodeError as e:  # Decode error
+            # Fallback to latin1
             target_io_size = READ_BLOCKSIZE
-            block_main = fp.read(target_io_size)
+            block_main = e.object.decode("latin1")
```

Comment:
I don't know how to handle this case to be honest. Is it particular to my setup?

Shall we have that to a list of possible encodings and try each of them until it fails?